src/core/controller/index.ts

```diff
@@ -404,6 +415,59 @@ export class Controller {
 		}
 	}
 
+	updateBackgroundCommandState(running: boolean, taskId?: string) {
+		const nextTaskId = running ? taskId : undefined
+		if (this.backgroundCommandRunning === running && this.backgroundCommandTaskId === nextTaskId) {
+			return
+		}
+		console.log(
+			"[Controller] updateBackgroundCommandState",
+			JSON.stringify({ running, taskId, previousTaskId: this.backgroundCommandTaskId }),
+		)
+		this.backgroundCommandRunning = running
+		this.backgroundCommandTaskId = nextTaskId
+		void this.postStateToWebview()
+	}
+
+	async cancelBackgroundCommand(): Promise<void> {
+		const didCancel = await this.task?.cancelBackgroundCommand()
+		if (!didCancel) {
+			this.updateBackgroundCommandState(false)
+		}
+	}
+
+	/**
+	 * Check if we should show the background terminal suggestion based on shell integration warning frequency
+	 * @returns true if we should show the suggestion, false otherwise
+	 */
+	shouldShowBackgroundTerminalSuggestion(): boolean {
```

Comment:
Should we check if the user has background terminal still enabled first before we show the suggestion?